lockfile/src/spdx.rs

```diff
@@ -76,62 +77,88 @@ fn type_from_url(url: &str) -> anyhow::Result<PackageType> {
     }
 }
 
+fn from_purl(pkg_url: &str, pkg_info: &PackageInformation) -> anyhow::Result<Package> {
+    let purl = PackageUrl::from_str(pkg_url)?;
+
+    let package_type = PackageType::from_str(purl.ty())
+        .or_else(|_| type_from_url(&pkg_info.download_location))
+        .context(UnknownEcosystem)?;
+
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
+
+    let pkg_version = pkg_info
+        .version_info
+        .as_ref()
+        .ok_or(purl.version())
+        .map_err(|_| anyhow!("No version found for `{}`", pkg_info.name))?;
+
+    let version = purl
+        .qualifiers()
+        .iter()
+        .find_map(|(key, value)| match key.as_ref() {
+            "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                version: pkg_version.clone(),
+                registry: value.to_string(),
+            })),
+            "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+            "vcs_url" => {
+                if value.as_ref().starts_with("git+") {
+                    Some(PackageVersion::Git(value.to_string()))
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        })
+        .unwrap_or(PackageVersion::FirstParty(pkg_version.into()));
+
+    Ok(Package { name, version, package_type })
+}
+
+fn from_locator(registry: &str, locator: &str) -> anyhow::Result<Package> {
+    let package_type = PackageType::from_str(registry).map_err(|_| UnknownEcosystem)?;
+    let (name, version) = match package_type {
+        PackageType::Npm => locator.rsplit_once('@'),
+        PackageType::Nuget => locator.rsplit_once('/'),
+        PackageType::Maven => locator.rsplit_once(':').filter(|(name, _)| name.contains(':')),
+        _ => {
+            // Not in the spec, but included for compatibility with our API
+            locator.rsplit_once('@')
+        },
+    }
+    .ok_or(anyhow!("Invalid locator: {}", locator))?;
+
+    let name = decode(name).map_err(|e| anyhow!("Error: {:?} when decoding {}", e, name))?;
```

Comment:
Don't throw away the underlying error. This should just add context

```suggestion
    let name = decode(name).with_context(|| anyhow!("URL decode failed: {:?}", name))?;
```